### DISCUSSION TITLE
How much time will skywalking slow down when application is starting?

### DISCUSSION THREAD

**Discussion by furaul** [apache/skywalking#9005] [Q&A] [answered]
I have an application, its starting costs about 230 seconds, when I use skywalking in it, its starting costs about 400 seconds, is this the situation expected, or my application is a special case?

My several other applications are also slow down about %50+, is there any metric data that how much time will skywalking slow down when application is starting?

I have profiled my application, and I found that most of time is spended on 
`net.bytebuddy.dynamic.ClassFileLocator.ForClassLoader#locate(java.lang.ClassLoader, java.lang.String)
java.lang.ClassLoader#getResourceAsStream` in bytebuddy.
Is this the situation expected?

My application is based on springboot, and it will load about 10000+ classes while starting.

**Comment by lujiajing1126** [comment]
The problem is actually out of the scope of the Skywalking Java Agent. As you said, this is related to the upstream module `bytebuddy`, with which the SW Java Agent uses to manipulate byte codes.

The time spent on the class file transformation actually depends on how many classes that should be transformed during the startup process. It could then be possibly affected by how many plugins you enable.

We've also observed the similar phenomena in both production and test environment. Since we deploy almost all application on the Kubernetes, the following measures are taken to alleviate the issue,

1. Increate CPU limit to about 1-1.5 cores per application (allow the application to use more cores during startup),
2. Delay initial detection time for readiness and liveness probes to over 90 seconds (tolerate more startup time)

**Comment by wu-sheng** [comment, marked as answer]
Booting performance is not our focus. If you would like to profile or optimize, we are open to discuss.

**Reply by wu-sheng** [reply]
Also, how many classes you want to instrument, that is your decision. A lot of plugins are optional and configurable.

**Reply by lujiajing1126** [reply]
Is that possible we introduce some metrics to track the class file transformation performance as a part of self observability.

**Reply by wu-sheng** [reply]
There are no kernel-level metrics in the agent system. If we want to add this, new self-observability APIs on the top of meter core should be established first.

**Reply by furaul** [reply]
I am researching on this topic, here is the issue I submitted on byte-buddy. https://github.com/raphw/byte-buddy/issues/1246

My train of thought is that maybe we could decrease the times of `java.lang.ClassLoader#getResourceAsStream` which would be called, or optimize the elapsed time of `java.lang.ClassLoader#getResourceAsStream`.

Another way is that could the enhancing of classes been processed after the booting of application, like alibaba-Arthas?

**Reply by wu-sheng** [reply]
Arthas provides attach mode, which is negative performance impact if we use this mode to do tracing.
Not all agents work in the same way.

**Reply by furaul** [reply]
> Arthas provides attach mode, which is negative performance impact if we use this mode to do tracing. Not all agents work in the same way.

I get it, thanks for the explain.
